modules/evm/src/lib.rs

```diff
@@ -467,21 +618,312 @@ impl<T: Trait> Module<T> {
+			ensure!(
+				account_info.storage_rent_deposit >= adjust_deposit,
+				Error::<T>::StorageQuotaNotEnough
+			);
+
+			account_info.storage_rent_deposit = account_info
+				.storage_rent_deposit
+				.checked_sub(&adjust_deposit)
+				.ok_or(Error::<T>::NumOutOfBound)?;
+			account_info.storage_quota = account_info
+				.storage_quota
+				.checked_sub(bytes)
+				.ok_or(Error::<T>::NumOutOfBound)?;
+
+			ensure!(
+				account_info.storage_usage <= account_info.storage_quota,
+				Error::<T>::StorageQuotaNotEnough
+			);
+
+			let additional_storage = {
+				let code_size = CodeInfos::get(contract_info.code_hash).map_or(0, |code_info| code_info.code_size);
+				contract_info
+					.total_storage_size()
+					.saturating_add(code_size)
+					.saturating_sub(account_info.storage_quota)
+			};
+
+			ensure!(additional_storage.is_zero(), Error::<T>::StorageQuotaNotEnough);
+			ensure!(
+				T::Currency::unreserve(&contract_info.maintainer, adjust_deposit).is_zero(),
+				Error::<T>::UnreserveFailed
+			);
+
+			Ok(())
+		})
+	}
+
+	fn do_request_transfer_maintainer(who: T::AccountId, contract: H160, new_maintainer: H160) -> DispatchResult {
+		Accounts::<T>::get(contract).map_or(Err(Error::<T>::ContractNotFound), |account_info| {
+			account_info
+				.contract_info
+				.map_or(Err(Error::<T>::ContractNotFound), |_| Ok(()))
+		})?;
+		ensure!(
+			PendingTransferMaintainers::<T>::get(contract, new_maintainer).is_none(),
+			Error::<T>::PendingTransferMaintainersExists
+		);
+
+		let transfer_maintainer_deposit = T::TransferMaintainerDeposit::get();
+		let to = T::AddressMapping::to_account(&contract);
+		T::Currency::transfer(&who, &to, transfer_maintainer_deposit, ExistenceRequirement::KeepAlive)?;
+		T::Currency::reserve(&to, transfer_maintainer_deposit)?;
+		PendingTransferMaintainers::<T>::insert(contract, new_maintainer, transfer_maintainer_deposit);
+		Ok(())
+	}
+
+	fn do_confirm_transfer_maintainer(who: T::AccountId, contract: H160, new_maintainer: H160) -> DispatchResult {
+		PendingTransferMaintainers::<T>::mutate_exists(
+			contract,
+			new_maintainer,
+			|maybe_transfer_maintainer_deposit| -> DispatchResult {
+				let transfer_maintainer_deposit = maybe_transfer_maintainer_deposit
+					.as_mut()
+					.ok_or(Error::<T>::PendingTransferMaintainersNotExists)?;
+
+				Accounts::<T>::mutate(contract, |maybe_account_info| -> DispatchResult {
+					let account_info = maybe_account_info.as_mut().ok_or(Error::<T>::ContractNotFound)?;
+					let contract_info = account_info
+						.contract_info
+						.as_mut()
+						.ok_or(Error::<T>::ContractNotFound)?;
+
+					ensure!(who == contract_info.maintainer, Error::<T>::NoPermission);
+
+					let maintainer = T::AddressMapping::to_account(&new_maintainer);
+
+					let from = T::AddressMapping::to_account(&contract);
+					ensure!(
+						T::Currency::unreserve(&from, *transfer_maintainer_deposit).is_zero(),
+						Error::<T>::UnreserveFailed
```

Comment:
I don't think we need the transfer and it is ok to ignore this unreserve error.

Reply:
https://github.com/paritytech/substrate/blob/master/frame/balances/src/lib.rs#L1148

unreserve amount maybe not the expected value